File: src/quest_editor/controllers/NpcCountsController.ts
import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { npc_data, NpcType } from "../../core/data_formats/parsing/quest/npc_types";
import { property } from "../../core/observable";

export type NameWithCount = { readonly name: string; readonly count: number };

export class NpcCountsController extends Controller {
    readonly npc_counts: Property<readonly NameWithCount[]>;
    readonly unavailable: Property<boolean>;

    constructor(store: QuestEditorStore) {
        super();

        this.unavailable = store.current_quest.map(q => q == undefined);

        this.npc_counts = store.current_quest
            .flat_map(quest => (quest ? quest.npcs : property([])))
            .map(this.update_view);
    }

    private update_view(npcs: readonly QuestNpcModel[]): NameWithCount[] {
        const npc_counts = new Map<NpcType, number>();

        for (const npc of npcs) {
            const val = npc_counts.get(npc.type) || 0;
            npc_counts.set(npc.type, val + 1);
        }

        const extra_canadines = (npc_counts.get(NpcType.Canane) || 0) * 8;

        // Sort by canonical order.
        const sorted_npc_counts = [...npc_counts].sort((a, b) => a[0] - b[0]);

        return sorted_npc_counts.map(([npc_type, count]) => {
            const extra = npc_type === NpcType.Canadine ? extra_canadines : 0;

            return { name: npc_data(npc_type).name, count: count + extra };
        });
    }
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/controllers/NpcCountsController.kt
package world.phantasmal.web.questEditor.controllers

import world.phantasmal.psolib.fileFormats.quest.NpcType
import world.phantasmal.cell.Cell
import world.phantasmal.cell.flatMap
import world.phantasmal.cell.isNull
import world.phantasmal.cell.list.emptyListCell
import world.phantasmal.cell.map
import world.phantasmal.web.questEditor.models.QuestNpcModel
import world.phantasmal.web.questEditor.stores.QuestEditorStore
import world.phantasmal.webui.controllers.Controller

class NpcCountsController(private val store: QuestEditorStore) : Controller() {
    val unavailable: Cell<Boolean> = store.currentQuest.isNull()

    val npcCounts: Cell<List<NameWithCount>> = store.currentQuest
        .flatMap { it?.npcs ?: emptyListCell() }
        .map(::countNpcs)

    fun focused() {
        store.makeMainUndoCurrent()
    }

    private fun countNpcs(npcs: List<QuestNpcModel>): List<NameWithCount> {
        val npcCounts = mutableMapOf<NpcType, Int>()
        var extraCanadines = 0

        for (npc in npcs) {
            // Don't count Vol Opt twice.
            if (npc.type != NpcType.VolOptPart2) {
                npcCounts[npc.type] = (npcCounts[npc.type] ?: 0) + 1

                // Cananes always come with 8 canadines.
                if (npc.type == NpcType.Canane) {
                    extraCanadines += 8
                }
            }
        }

        return npcCounts.entries
            // Sort by canonical order.
            .sortedBy { (npcType) -> npcType.ordinal }
            .map { (npcType, count) ->
                val extra = if (npcType == NpcType.Canadine) extraCanadines else 0
                NameWithCount(npcType.simpleName, (count + extra).toString())
            }
    }

    data class NameWithCount(val name: String, val count: String)
}
